ProcessTree.h

```diff
@@ -0,0 +1,142 @@
+/*
+    microsoft-oms-auditd-plugin
+
+    Copyright (c) Microsoft Corporation
+
+    All rights reserved.
+
+    MIT License
+
+    Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the ""Software""), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef AUOMS_PROCESSTREE_H
+#define AUOMS_PROCESSTREE_H
+
+#include "RunBase.h"
+#include "UserDB.h"
+#include "ProcessDefines.h"
+#include "FiltersEngine.h"
+
+#include <string>
+#include <unordered_map>
+#include <queue>
+#include <chrono>
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+#include <sys/types.h>
+#include <unistd.h>
+#include <sys/socket.h>
+#include <linux/netlink.h>
+#include <linux/connector.h>
+#include <linux/cn_proc.h>
+#include <bits/stdc++.h>
+
+enum ProcessTreeSource { ProcessTreeSource_execve, ProcessTreeSource_pnotify, ProcessTreeSource_procfs };
+
+class FiltersEngine;
+
+struct Ancestor {
+    int pid;
+    std::string exe;
+};
+
+class ProcessTreeItem {
+public:
+    ProcessTreeItem(enum ProcessTreeSource source, int pid, int ppid=0):
+        _source(source), _pid(pid), _ppid(ppid), _uid(-1), _gid(-1), _flags(0), _exec_propagation(0), _exited(false) {}
+    ProcessTreeItem(enum ProcessTreeSource source, int pid, int ppid, int uid, int gid, const std::string& exe, const std::string& cmdline):
+        _source(source), _pid(pid), _ppid(ppid), _uid(uid), _gid(gid), _exe(exe), _cmdline(cmdline),
+        _flags(0), _exec_propagation(0), _exited(false) {}
+
+    enum ProcessTreeSource _source;
+    int _pid;
+    int _ppid;
+    int _uid;
+    int _gid;
+    std::vector<int> _children;
```

Comment:
_children is used for process information propagation.  This allows for fork and exec notifications and execve events to arrive in any order and for them to still annotate the new processes with the correct information. Without this mechanism, some processes will be created with incorrect information and will be incorrectly assessed for filtering. The _children vectors allow for process information to propagate correctly to avoid these errors.